x-pack/plugins/cases/server/authorization/authorization.ts

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { KibanaRequest } from 'kibana/server';
+import Boom from '@hapi/boom';
+import { SecurityPluginStart } from '../../../security/server';
+import { PluginStartContract as FeaturesPluginStart } from '../../../features/server';
+import { AuthorizationFilter, GetSpaceFn } from './types';
+import { getOwnersFilter } from './utils';
+import { AuthorizationAuditLogger, OperationDetails } from '.';
+
+/**
+ * This class handles ensuring that the user making a request has the correct permissions
+ * for the API request.
+ */
+export class Authorization {
+  private readonly request: KibanaRequest;
+  private readonly securityAuth: SecurityPluginStart['authz'] | undefined;
+  private readonly featureCaseOwners: Set<string>;
+  private readonly auditLogger: AuthorizationAuditLogger;
+
+  private constructor({
+    request,
+    securityAuth,
+    caseOwners,
+    auditLogger,
+  }: {
+    request: KibanaRequest;
+    securityAuth?: SecurityPluginStart['authz'];
+    caseOwners: Set<string>;
+    auditLogger: AuthorizationAuditLogger;
+  }) {
+    this.request = request;
+    this.securityAuth = securityAuth;
+    this.featureCaseOwners = caseOwners;
+    this.auditLogger = auditLogger;
+  }
+
+  /**
+   * Creates an Authorization object.
+   */
+  static async create({
+    request,
+    securityAuth,
+    getSpace,
+    features,
+    auditLogger,
+  }: {
+    request: KibanaRequest;
+    securityAuth?: SecurityPluginStart['authz'];
+    getSpace: GetSpaceFn;
+    features: FeaturesPluginStart;
+    auditLogger: AuthorizationAuditLogger;
+  }): Promise<Authorization> {
+    // Since we need to do async operations, this static method handles that before creating the Auth class
+    let caseOwners: Set<string>;
+    try {
+      const disabledFeatures = new Set((await getSpace(request))?.disabledFeatures ?? []);
+
+      caseOwners = new Set(
+        features
+          .getKibanaFeatures()
+          // get all the features' cases owners that aren't disabled
+          .filter(({ id }) => !disabledFeatures.has(id))
+          .flatMap((feature) => feature.cases ?? [])
+      );
+    } catch (error) {
+      caseOwners = new Set<string>();
+    }
```

Comment:
I think the reason we wrap it in a `try/catch` is for the get spaces call since it's an `async` function.

> Instead of continuing, should we instead be throwing an error and capturing the root cause?

Yeah that sounds good to me, should we just log a message and re-throw, or should we do an audit logger message and re-throw?

Reply:
I don't think this is an event worth auditing. I would just log the message and re-throw